### ISSUE TITLE
[FEA] Support StringColumn equality comparisons to a single string

### ISSUE THREAD

**Issue by beckernick** [rapidsai/cudf#1808]
**Describe the bug**
#1337 included support for operations like `df['a'] == df['b']` (though it's currently broken -- #1750 ) but not comparisons to a single string. We need to support binary ops like equality on StringColumns to allow series/dataframe indexing based on direct equality comparisons. Doing something like this causes an error because our binary ops do not effectively support strings (only NumericalColumns have `normalize_binop_value`):

```python
import cudf

df = cudf.DataFrame({'a':['a','b','c','d'], 'b':[0,1,2,3]})
df.a == 'b'
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-253-b0751be328af> in <module>
      2 
      3 df = cudf.DataFrame({'a':['a','b','c','d'], 'b':['b','c','d','e']})
----> 4 df.a == 'b'

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.8.0a1+47.gd6b7bd1b.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/series.py in __eq__(self, other)
    558 
    559     def __eq__(self, other):
--> 560         return self._unordered_compare(other, 'eq')
    561 
    562     def equals(self, other):

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.8.0a1+47.gd6b7bd1b.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/series.py in _unordered_compare(self, other, cmpops)
    541     def _unordered_compare(self, other, cmpops):
    542         nvtx_range_push("CUDF_UNORDERED_COMP", "orange")
--> 543         other = self._normalize_binop_value(other)
    544         outcol = self._column.unordered_compare(cmpops, other._column)
    545         result = self._copy_construct(data=outcol)

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.8.0a1+47.gd6b7bd1b.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/series.py in _normalize_binop_value(self, other)
    536             return Series(other)
    537         else:
--> 538             col = self._column.normalize_binop_value(other)
    539             return self._copy_construct(data=col)
    540 

AttributeError: 'StringColumn' object has no attribute 'normalize_binop_value'
```

One hack-y solution leveraging `nvstrings.compare` is something like the following:

```python
import cudf

df = cudf.DataFrame({'a':['a','b','c','d'], 'b':[0,1,2,3]})

comparison_val = 'c'

arr = np.arange(df.a.data.size(),dtype=np.int32)
d_arr = librmm.to_device(arr)
s.compare(comparison_val, devptr=d_arr.device_ctypes_pointer.value)
mask = cudf.Series(d_arr) # could use a column here instead
boolmask = mask == 0

print(df[boolmask])
   a  b
2  c  2
```

But this isn't sustainable.

**Comment by beckernick** [comment]
This could likely be resolved by simply adding a basic `normalize_binop_value` to the StringColumn class as simple as

```python
def normalize_binop_value(self, other):
        if other_dtype.kind in 'U':
            return self
        else:
            raise TypeError('cannot broadcast {}'.format(type(other)))
```

The results will be all `True` until #1750 is fixed, though.

@randerzander

**Comment by shwina** [comment]
Here's an implementation of `normalize_binop_value` that should allow this (and also some other nice things):

https://github.com/rapidsai/cudf/pull/1807/files#diff-311cc94833eaeee5562fe39544aa62abR583

@beckernick If needed I can break that piece away and put in a separate PR if it looks good to you.

**Comment by beckernick** [comment]
@shwina , since your improved `normalize_binop_value` will not resolve the core user issue (as the real functionality is ultimately blocked by #1750), don't feel any need to do that as a result of this issue.

With that said, in general I think it's valuable for large PRs composed of somewhat stand-alone / weakly coupled functionality to be split up into smaller PRs. It makes them easier to review and allows uncoupled functionality to be merged in the event that a small portion of the large PR is thornier than expected.

Defer to you and Keith on whether that makes sense for #1807 or not. If you keep #1807 as is, please note in the summary that it will close #1808 👍

**Comment by kkraus14** [comment]
@devavret was this fixed in #1891?

**Comment by devavret** [comment]
No, this needs `gdf_scalar` to support string values. And that needs some discussion on how to do it. Do we also add `dtype_info` to `gdf_scalar`? And do we store an NVCategory object in it? Or do we store just a `char*` in the data union or as a separate member of `gdf_scalar`

**Comment by harrism** [comment]
Devavret's point is exactly why I deferred supporting `cudf::fill()` for scalar string values in #1908 . So similar to comparing StringColumns to a single string, I think today we can't assign a single index or a range of indices to a single string.

**Comment by kkraus14** [comment]
@devavret @harrism @shwina If we're only looking at supporting equality and inequality here, wouldn't it be easy enough to just get the integer encoding of the string scalar and use that for comparison? I.E. if the string isn't in the category than equals is always False, not equals is always True. If the string is in the category then we get the integer index of it via a `get_value` call and then use that for a binaryop (in)equality comparison.

**Comment by harrism** [comment]
That's true, that should work. It won't work for assignment, but will for comparison.

**Comment by beckernick** [comment]
@harrism @devavret @kkraus14 @shwina this now works in Python land with Ashwin's updated normalize_binop_value. Specifically, due to this I believe:

https://github.com/rapidsai/cudf/blob/367dcf3d338536f8288a24610b3300be32d43e3e/python/cudf/cudf/dataframe/string.py#L754-L757

We can leave this issue open if it's still not resolved on the libcudf side, but from a Python user's perspective this is now functional.

**Comment by kkraus14** [comment]
I think we should close this as we already have an issue open about supporting strings in gdf_scalar.

**Comment by beckernick** [comment]
Sounds good, closing.
